### PR TITLE
Backward operation of torch.eig for real eigenvalues

### PR DESCRIPTION
**Pull request by mfkasim1**
Another pull request to follow up issue #32531. 
Here I implemented the backward operation for `torch.eig` with a condition that all the eigenvalues are real.

This pull request is independent of my another pull request #32932, which means that there is no dependency between this PR and my another PR.

### CODE REVIEW

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+  // This gradient only works for real eigenvalues at the moment.
+  TORCH_CHECK(eigenvectors,
+           "eig_backward: Setting eigenvectors to false in torch.eig doesn't compute eigenvectors ",
+           "and hence we cannot compute backward. Please use torch.eig(eigenvectors=True)");
+  auto imag_idx = at::ones({1}, lambda.options().dtype(at::kLong));
+  auto real_idx = at::zeros({1}, lambda.options().dtype(at::kLong));
+  auto zeros = at::zeros({1}, lambda.options());
+  TORCH_CHECK(
+      at::allclose(
+        at::index_select(lambda, /*dim=*/-1, /*index=*/imag_idx),
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: If this is only selecting the `1`th entry in the last dimension, a more efficient way to do it is with `lambda.index(-1, 1)`.
`index_select()` copies the values while `index()` does not.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+      "eig_backward: Backward calculation does not support complex eigenvalues at the moment.");
+
+  auto glambda = grads[0];
+  auto gv = grads[1];
+  auto vt = v.transpose(-2, -1);
+
+  Tensor result;
+  // contribution from the eigenvectors
+  if (gv.defined()) {
+    auto rlambda = at::index_select(lambda, /*dim=*/-1, /*index=*/real_idx);
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: Same here, `lambda.index(-1, 0)` will be faster and avoid the copy.

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    hm.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+    hm.pow_(-1.0);
+
+    auto gvortho = gv - at::sum(gv * v, /*dim=*/-2, /*keepdim=*/true) * v;
+    auto B = hm * at::matmul(vt, gvortho);
+    auto A = at::matmul(B, vt);
+
+    std::tie(result, std::ignore) = at::solve(A, vt);
+  }
+  else {
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: nit: extra new-line: `} else {`.

File: tools/autograd/templates/Functions.cpp [hunk 4]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    result = at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  }
+  // contribution from eigenvalues
+  if (glambda.defined()) {
+    Tensor result1;
+    auto grlambda = at::index_select(glambda, /*dim=*/-1, /*index=*/real_idx) * vt;
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: `index_select` -> `select`

File: tools/autograd/templates/Functions.cpp [hunk 5]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    auto A = at::matmul(v, grlambda);
+    auto vvt = at::matmul(v, vt);
+    std::tie(result1, std::ignore) = at::solve(A, vvt);
+    result = result.add(result1);
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: This will only work if result is already defined. You want to protect it with `if (result.defined())` and otherwise just do the assignment.

File: tools/autograd/templates/Functions.cpp [hunk 6]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
   return u_term + sigma_term + v_term;
 }
 
+Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: Do you have a reference containing the gradient formula?
That can come in handy if we come back to this later.

File: test/test_autograd.py [hunk 7]
```diff
@@ -2299,6 +2299,29 @@ def func(root, b, upper):
...
+            A = random_symmetric_matrix(dims[-1], *dims[:-2])
+            U = torch.rand(*dims)
+            Uinv = torch.inverse(U)
+            B = torch.matmul(Uinv, torch.matmul(A, U)).requires_grad_()
+
+            gradcheck(func, [B])
+            gradgradcheck(func, [B])
+
+        for dims in [(3, 3), (5, 5)]:
+            run_test(dims)
```
Comment (albanD) [inline test/test_autograd.py]: Could you add a test here where only one of the two outputs sends gradients?
In particular, to make sure the logic if one grad are not defined is working fine.

**Review by albanD — COMMENTED** [review]
This looks quite good!
Small details about test and perf but this is almost ready to go.

Will this need an update when the batch eig lands? Or are these formulas going to work for both?

File: tools/autograd/templates/Functions.cpp (hunk 1 above)
Comment (mfkasim91) [reply tools/autograd/templates/Functions.cpp]: `.index()` seems to be accepting list of tensor index and the documentation does not seem to be clear, so I use `.slice()`. Hopefully `slice()` does not make a copy.

File: tools/autograd/templates/Functions.cpp (hunk 3 above)
Comment (mfkasim91) [reply tools/autograd/templates/Functions.cpp]: Done

File: tools/autograd/templates/Functions.cpp (hunk 5 above)
Comment (mfkasim91) [reply tools/autograd/templates/Functions.cpp]: The result has already been defined in the previous block. If `gv.defined()`, then result is defined as one of the output of `solve()`. Otherwise, it is defined as `zeros_like(self)`.

The idea comes from `symeig_backward` which uses the similar strategy.

File: tools/autograd/templates/Functions.cpp (hunk 6 above)
Comment (mfkasim91) [reply tools/autograd/templates/Functions.cpp]: I derived it myself, but there is a document that contains similar expression (and I have put a link to it). However, the expression might be slightly different, though.

File: test/test_autograd.py (hunk 7 above)
Comment (mfkasim91) [reply test/test_autograd.py]: Sure, added them in the new commit.

**Comment by mfkasim91** [comment]
Thanks for the review, @albanD. The computation in this PR should not be changed when my another PR lands. The way I wrote it is by following closely `symeig_backward` which should be fine with batched calculation.
The only thing to change when `eig` is batched is the test case which should include tensor > 2D.

File: tools/autograd/templates/Functions.cpp (hunk 5 above)
Comment (albanD) [reply tools/autograd/templates/Functions.cpp]: Ho right. But then the `else` part of the previous block is not needed. It create a Tensor full of zeros that is not necessary :)

Comment (mfkasim91) [reply tools/autograd/templates/Functions.cpp]: Yes, you're right. I've changed it in the new commit.

**Review by albanD — APPROVED** [review]
Looks good, thanks !

File: tools/autograd/templates/Functions.cpp [hunk 8]
```diff
@@ -1792,6 +1792,55 @@ Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+  // contribution from the eigenvectors
+  if (gv.defined()) {
+    auto rlambda = lambda.slice(/*dim=*/-1, /*start=*/0, /*end=*/1);
+
+    auto hm = rlambda.transpose(-2,-1) - rlambda;
+    hm.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(INFINITY);
+    hm.pow_(-1.0);
+
+    auto gvortho = gv - at::sum(gv * v, /*dim=*/-2, /*keepdim=*/true) * v;
+    auto B = hm * at::matmul(vt, gvortho);
```
Comment (nikitaved) [inline tools/autograd/templates/Functions.cpp:1822]: @mfkasim1 , could you please explain this line? It does not seem to agree with the reference you mentioned above.

Comment (mfkasim1) [reply tools/autograd/templates/Functions.cpp:1822]: I actually derived the expression myself, but I didn't have the time to write it down, so I just put the closest available literature in the code.

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:1822]: The derivations are different, so I do not know which method to trust, but I do not immediately see any issues in the reference. So, then, how did you get that `gvortho`?

Comment (mfkasim1) [reply tools/autograd/templates/Functions.cpp:1822]: I'm not sure if they are actually different or actually the same with just different expressions.
I have put my expression into `gradcheck` and `gradgradcheck` and it passes.
I will write the derivation down if I got time.

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:1822]: They are different, in your derivation you subtract from the `gv` the projection of `gv` into `v`, and they do not do it in the reference.

Comment (mfkasim1) [reply tools/autograd/templates/Functions.cpp:1822]: Ah, if I remember it correctly, I subtract it because of the constraint of the eigenvectors to have `norm = 1`.
I probably did the similar trick from [this paper](https://arxiv.org/abs/2011.04366) (eq. 3.23-3.25)

Comment (nikitaved) [reply tools/autograd/templates/Functions.cpp:1822]: All right, now I see. I did check the output of `torch.eig` it indeed produces normalized eigenvectors. Awesome, thank you!

**Comment by kimjw0623** [comment]
Is this feature updated to pytorch 1.7.1?

**Comment by mfkasim1** [comment]
@kimjw0623 yes, it should be. You can check.
